csharp/src/Google.Protobuf/Reflection/Dynamic/DynamicMessage.cs

```diff
@@ -0,0 +1,377 @@
+﻿using Google.Protobuf.Collections;
+using System;
+using System.Collections;
+
+namespace Google.Protobuf.Reflection.Dynamic
+{
+
+    /// <summary>
+    /// An implementation of IMessage that can represent arbitrary types, given a MessageaDescriptor.
+    /// Unknown fields not supported.
+    /// </summary>
+    public sealed partial class DynamicMessage : IMessage
+    {
+        private readonly MessageDescriptor type;
+        private readonly FieldSet fields;
+        private int memoizedSize = -1;
+
+        /// <summary>
+        /// Descriptor for this message.
+        /// </summary>
+        public MessageDescriptor Descriptor => type;
+
+        /// <summary>
+        /// Default implementation of Overridden method.
+        /// </summary>
+        /// <param name="input"></param>
+        /// <exception cref="NotImplementedException"></exception>
+        public void MergeFrom(CodedInputStream input)
+        {
+            throw new NotImplementedException();
+        }
+        /// <summary>
+        /// Writes the message to CodedOutputStream.
+        /// </summary>
+        /// <param name="output"></param>
+        public void WriteTo(CodedOutputStream output)
+        {
+            fields.WriteTo(output);
+        }
+
+        /// <summary>
+        /// Used to calculate the size of the serialized message.
+        /// </summary>
+        /// <returns></returns>
+        public int CalculateSize()
+        {
+            int size = memoizedSize;
+            if (size != -1)
+            {
+                return size;
+            }
+            size = fields.GetSerializedSize();
+            memoizedSize = size;
+            return size;
+        }
+
+        /// <summary>
+        /// Parse data as a message of the given type and return it.
+        /// </summary>
+        /// <param name="type"></param>
+        /// <param name="data"></param>
+        /// <returns></returns>
+        public static DynamicMessage ParseFrom(MessageDescriptor type, ByteString data)
+        {
+            Builder builder = NewBuilder(type);
+            builder.MergeFrom(data);
+            return builder.BuildParsed();
+        }
+
+        /// <summary>
+        /// Construct a IMessage.Builder for the given type.
+        /// </summary>
+        /// <param name="type"></param>
+        /// <returns></returns>
+        public static Builder NewBuilder(MessageDescriptor type)
+        {
+            return new Builder(type);
+        }
+
+        /// <summary>
+        /// Returns the value of the field, represented by field descriptor
+        /// </summary>
+        /// <param name="fd"></param>
+        /// <returns></returns>
+        public object GetField(FieldDescriptor fd)
+        {
+            return fields.GetField(fd);
+        }
+
+        private DynamicMessage(MessageDescriptor type, FieldSet fields)
```

Comment:
Constructors at the top. +there's an extra new line after this method

Reply:
Done